### ISSUE TITLE
False positive type annotation for two different `Callable`

### ISSUE THREAD

**Issue by Fanteria** [microsoft/pyright#4007]
**Describe the bug**
An error is shown when the argument has type annotation with two different `Callable`. If only one `Callable` is used, then one error disappears and the second stay as expected.

**To Reproduce**
The minimal example below shows the error for me. Comments in the minimal example represent errors.

**Expected behavior**
I think it is a false positive error and should not be shown.

**Screenshots or Code**
Minimal example:
```python
from inspect import signature
from collections.abc import Callable


class A:
    def oneArg(self) -> None:
        print("Only one argument")

    def twoArgs(self, name: str) -> None:
        print(f"Hello {name}, two arguments here.")


class B:
    def __init__(self) -> None:
        self.a = A()

    def printFunction(
        self, toPrint: Callable[[A], None] | Callable[[A, str], None]
    ) -> None:
        if len(signature(toPrint).parameters) == 1:
            toPrint(self.a)  # Expected 1 more positional argument
        else:
            toPrint(self.a, "John")  # Expected 1 positional argument


b = B()
b.printFunction(A.oneArg)
b.printFunction(A.twoArgs)
```

**VS Code extension or command-line**
Command-line as a server for LSP in Neovim.

**Additional context**
Pyright version 1.1.251
Python version 3.10.7

**Comment by erictraut** [comment]
Yes, pyright doesn't support the expression form `len(signature(x).parameters) == y` for type narrowing. For a full list of supported type guard patterns, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards).

Each new type guard pattern requires custom logic that must be implemented and maintained. We do occasionally add support for new patterns but only in cases where we have evidence that it's a commonly-used pattern. I've never seen this particular pattern used before, and I think it's quite uncommon. As further evidence that it's not common, mypy also doesn't support this type narrowing pattern, and it consequently emits the same errors as pyright in this case.

I'm going to close this feature request. It's possible that we might reconsider it in the future if the issue receives a sufficient number of upvotes (in the form of thumbs-up reactions), but I think it's unlikely. This is a very complex expression form that uses very specific function calls (`len` and `inspect.signature`) and type attributes (`parameters`).

If you want this code to type check in pyright (or mypy, for that matter), you'll need to find a different approach.

**Comment by lekcyjna123** [comment]
Hi,
I had the same problem and I was able to workaround it using custom type guard:
```py
T1 = TypeVar("T1")
T2 = TypeVar("T2")
T3 = TypeVar("T3")

def is_one_arg_callable(func: Callable[[T1], T2] | Callable[[Any, Any], Any]) -> TypeGuard[Callable[[T1], T2]]:
    return len(signature(func).parameters) == 1

def is_two_arg_callable(func: Union[Callable[[Any], Any], Callable[[T1, T2], T3]]) -> TypeGuard[Callable[[T1, T2], T3]]:
    return len(signature(func).parameters) == 2
```
Calling one of these two functions and checking result in if, before calling `func` fiz the problem.
